model/src/DA/Finance/Asset/Account.daml

```diff
@@ -0,0 +1,162 @@
+        exercise self Debit with depositCid; nominee >>= \asset -> exerciseByKey @AccountRule toAccount.id Credit with asset; nominee
+
+    nonconsuming choice Credit : ContractId AssetDeposit
+      -- ^ Credit an account.
+      with
+        asset : Asset
+        nominee : Party
+      controller account.owner, nominee
+      do
+        checkNominee nominee this
+
+        create AssetDeposit with account; asset; lock = None; observers
+
+    nonconsuming choice Debit : Asset
+      -- ^ Debit an account. The asset deposit to be debited is consumed in its
+      -- entirety.
+      with
+        depositCid : ContractId AssetDeposit
+        nominee : Party
+      controller account.owner, nominee
+      do
+        checkNominee nominee this
+
+        deposit <- fetchAndArchive depositCid
+        checkDepositAccount deposit this *> checkDepositLock deposit
+        pure deposit.asset
+
+    nonconsuming choice LockAssetDeposit : ContractId AssetDeposit
+      -- ^ Lock an asset deposit. Locking an asset deposit will block the asset
+      -- deposit from any action until the lock is released by one of the
+      -- parties set on the lock.
+      with
+        newLock : Lock
+        depositCid : ContractId AssetDeposit
+        nominee : Party
+      controller account.owner, nominee
+      do
+        checkNominee nominee this
+
+        deposit <- fetchAndArchive depositCid
+        checkDepositAccount deposit this *> checkDepositLock deposit
+        create deposit with lock = Some newLock
+
+    nonconsuming choice UnlockAssetDeposit : ContractId AssetDeposit
+      -- ^ Unlock an asset deposit. Requires the signature from one of the
+      -- `releasers` party set on the lock.
+      with
+        depositCid : ContractId AssetDeposit
+        releaser : Party
+      controller account.owner, releaser
+      do
+        deposit <- fetchAndArchive depositCid
+
+        case deposit.lock of
+          None -> abort "Account is not locked"
+          Some l -> do
+            assertMsg ("Party " <> show releaser <> " does not have the authority to unlock this account.") $ releaser `member` l.releasers
+            create deposit with lock = None
+
+    choice AddNominee : ContractId AccountRule
+      -- ^ Add a nominee to an account.
+      with
+        nominee : Party
+      controller account.owner, nominee
+      do
+        create this with nominees = nominee `insert` nominees
+
+    choice RemoveNominee : ContractId AccountRule
+      -- ^ Remove a nominee from an account. Requires the signature of the nominee
+      -- to ensure any pending transactions are completed or moved to an appropriate
+      -- state.
+      with
+        nominee : Party
+      controller account.owner, nominee
+      do
+        create this with nominees = nominee `delete` nominees
+
+    choice AddObserver : (ContractId AccountRule, [ContractId AssetDeposit])
+      -- ^ Add an observer to an account. Requires updating the observers of asset
```

Comment:
what's the intended use case for this choice? it's quite a "heavy" operation, having to update all asset deposits and we have no guarantee that the user does in fact pass in _all_ the deposits. so i'm kinda questioning the usefulness of having separate observers, maybe the parties currently involved (signatories, nominees, owner) are sufficient. but if there's a case where we have to make an asset deposit (resp. an account rule) visible to a 3rd party I'm curious to understand that.

Reply:
This is something I've copied over from the `AssetSettlementRule` with the exception that I've added logic to update existing `AssetDeposit`'s observers. The use case I'm thinking of is when you would like a party such as a spouse, a portfolio manager, an auditor, etc. access to your account and hence your positions. Is this a case we should cater for? I'm not sure, but I've added the logic so we can at least discuss it.